app/queries/get_application_choices_for_providers.rb

```diff
@@ -16,11 +16,29 @@ def self.call(providers:, vendor_api: false, includes: DEFAULT_INCLUDES)
 
     statuses = vendor_api ? ApplicationStateChange.states_visible_to_provider_without_deferred : ApplicationStateChange.states_visible_to_provider
 
-    ApplicationChoice.includes(*includes)
-      .where('courses.provider_id' => providers, 'courses.recruitment_cycle_year' => RecruitmentCycle.years_visible_to_providers)
-      .or(ApplicationChoice.includes(*includes)
-        .where('courses.accredited_provider_id' => providers, 'courses.recruitment_cycle_year' => RecruitmentCycle.years_visible_to_providers))
-      .where('status IN (?)', statuses)
-      .select('"application_choices"."id" AS id')
+    provider_ids_string = providers.map(&:id).join(', ')
+
+    ApplicationChoice.includes(*includes).from <<~GET_APPLICATION_CHOICES_FOR_PROVIDERS.squish
+      (
+        SELECT ac.*
+          FROM application_choices ac
+          INNER JOIN course_options co
+                  ON co.id = COALESCE(ac.offered_course_option_id, ac.course_option_id)
+          INNER JOIN courses c
+                  ON co.course_id = c.id
+          LEFT OUTER JOIN course_options original_co
+                  ON original_co.id = ac.course_option_id
+          LEFT OUTER JOIN courses original_c
+                  ON original_co.course_id = original_c.id
+          WHERE c.recruitment_cycle_year IN (#{RecruitmentCycle.years_visible_to_providers.join(', ')})
+            AND ac.status IN (#{statuses.map { |s| "'#{s}'" }.join(', ')})
+            AND (
+              c.provider_id IN (#{provider_ids_string})
+              OR c.accredited_provider_id IN (#{provider_ids_string})
+              OR original_c.provider_id IN (#{provider_ids_string})
+              OR original_c.accredited_provider_id IN (#{provider_ids_string})
+            )
+      ) AS application_choices
+    GET_APPLICATION_CHOICES_FOR_PROVIDERS
```

Comment:
As discussed, hopefully we can go SQL-free by using the AR statement from https://github.com/DFE-Digital/apply-for-teacher-training/compare/audits-for-application-choice-query-experiment . Moving the includes to the end of the statement keeps the `to_sql` cleaner. (Note that this broke a test, but the fix is on that branch too :) )

Reply:
Yes, having `applications.includes(*includes)` at the end of the chain makes a world of difference. I've revised  `GetApplicationChoicesForProviders` to reduce the amount of sql (but only just, to cater for the application visibility adjustments).